File: src/etch/bindings/java/transport/AbstractDeliveryService.java
package etch.bindings.java.transport;

import java.util.concurrent.TimeoutException;

import metreos.core.Who;
import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.Message;
import etch.bindings.java.msg.Type;
import etch.bindings.java.support.DeliveryService;
import etch.bindings.java.support.Mailbox;
import etch.bindings.java.support.StubBase;

/**
 * Abstraction of DeliveryService functionality.
 */
abstract public class AbstractDeliveryService
	implements DeliveryService, MessageHandler<MailboxMessageSource>
{
	public MailboxMessageSource getSource()
	{
		return src;
	}

	public void setSource( MailboxMessageSource src )
	{
		this.src = src;
	}

	private MailboxMessageSource src;

	/**
	 * Adds a stub to the set of stubs.
	 * @param stb
	 */
	public void addStub( StubBase stb )
	{
		if (this.stub != null)
			throw new UnsupportedOperationException( "only one stub for now" );
		this.stub = stb;
	}
	
	/**
	 * Removes all the stubs in the set of stubs.
	 */
	public void removeStub()
	{
		stub = null;
	}
	
	private StubBase stub;

	public Object sessionQuery( Object query )
	{
		if (stub != null)
			return stub.sessionQuery( query );
		
		throw new UnsupportedOperationException( "unknown query "+query );
	}

	public void sessionControl( Object control, Object value )
	{
		if (stub != null)
			stub.sessionControl( control, value );
		
		throw new UnsupportedOperationException( "unknown control "+control );
	}

	public void sessionNotify( Object event )
	{
		if (stub != null)
			stub.sessionNotify( event );
		else if (event instanceof Throwable)
			((Throwable) event).printStackTrace();
	}

	public boolean message( Who sender, Message msg ) throws Exception
	{
		return stub.message( sender, msg );
	}
	
	public Object transportQuery( Object query ) throws Exception
	{
		return src.transportQuery( query );
	}
	
	public void transportControl( Object control, Object value ) throws Exception
	{
		src.transportControl( control, value );
	}

	public void transportNotify( Object event ) throws Exception
	{
		src.transportNotify( event );
	}
	
	public void send( Message msg ) throws Exception
	{
		src.messagex( null, msg );
	}

	public Mailbox begincall( Message msg )
	{
		try
		{
			return src.message( null, msg, 1, 0 );
		}
		catch ( Exception e )
		{
			throw new RuntimeException( "unexpected exception sending message", e );
		}
	}

	public void endvoidcall( Mailbox mb, Type responseType,
		Field responseField, int timeout ) throws Exception
	{
		endcall( mb, responseType, responseField, timeout );
	}

	public Object endcall( Mailbox mb, Type responseType, Field responseField, int timeout ) throws Exception
	{
		try
		{
			Mailbox.Element mbe = mb.read( timeout );
			if (mbe == null)
				throw new TimeoutException( "timeout waiting for "+responseType );
			Message rmsg = mbe.msg;
			rmsg.checkType( responseType );
			Object r = rmsg.get( responseField );
			if (r instanceof Exception)
			{
				Exception e = (Exception) r;
				e.fillInStackTrace();
				throw e;
			}
			return r;
		}
		finally
		{
			mb.closeDelivery();
		}
	}
	
	@Override
	abstract public String toString();
}

File: src/etch/bindings/csharp/transport/AbstractDeliveryService.cs
using System;
using System.Collections.Generic;
using System.Text;
using Etch.Support;
using Etch.Msg;

namespace Etch.Transport
{
    public abstract class AbstractDeliveryService : DeliveryService, MessageHandler<MailboxMessageSource>
    {

        private MailboxMessageSource _src;

        public MailboxMessageSource GetSource()
        {
            return _src;
        }

        public void SetSource( MailboxMessageSource src )
        {
            _src = src; 
        }

        private StubBase _stub;

        /// <summary>
        /// Adds a stub to a list of stub
        /// </summary>
        /// <param name="stb"></param>
        public void AddStub( StubBase stb )
        {
            if ( _stub != null )
                throw new Exception( "Unsupported -- only one stub for now" );
            _stub = stb;
        }

        /// <summary>
        /// Removes all the stubs in the set of stubs
        /// </summary>
        public void RemoveStub()
        {
            _stub = null;
        }

        public Object SessionQuery( Object query )
        {
            if ( _stub != null )
                return _stub.SessionQuery( query );

            throw new Exception( "Unknown query " + query );
        }

        public void SessionControl( Object control, Object value )
        {
            if ( _stub != null )
                _stub.SessionControl( control, value );

            throw new Exception( "Unknown control " + control );
        }

        public void SessionNotify( Object eventObj )
        {
            if ( _stub != null )
                _stub.SessionNotify( eventObj );
            else if ( eventObj is Exception )
                Console.WriteLine( ( ( Exception ) eventObj ).StackTrace );
        }

        public Boolean Message( Who sender, Message msg )
        {
            return _stub.Message( sender, msg );
        }

        public Object TransportQuery( Object query )
        {
            return _src.TransportQuery( query );
        }

        public void TransportControl( Object control, Object value )
        {
            _src.TransportControl( control, value );
        }

        public void TransportNotify( Object eventObj )
        {
            _src.TransportNotify( eventObj );
        }

        public void Send( Message msg )
        {
            _src.Messagex( null, msg );
        }

        public Mailbox BeginCall( Message msg )
        {
            try
            {
                return _src.Message( null, msg, 1, 0 );
            }
            catch ( Exception e )
            {
                throw new Exception( "unexpected exception sending message", e );
            }
        }

        public Object EndCall( Mailbox mb, XType responseType, Field responseField,
            int timeout )
        {
            try
            {
                Element mbe = mb.Read( timeout );
                if ( mbe == null )
                    throw new TimeoutException( "timeout waiting for " + responseType );
                Message rmsg = mbe.msg;
                rmsg.CheckType( responseType );
                Object r = rmsg.Get( responseField );
                if ( r is Exception )
                {
                    Exception e = ( Exception ) r;
                    throw e;
                }
                return r;
            }
            finally
            {
                mb.CloseDelivery();
            }
        }

        public abstract override string ToString();
    }
}
